src/contextmenu.js

```diff
@@ -0,0 +1,472 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Multiple selection context menu.
+ */
+
+import * as Blockly from 'blockly/core';
+import {blockSelection, hasSelectedParent} from './global';
+
+/**
+ * Modification for context menu 'Duplicate' to be available for
+ * multiple blocks.
+ */
+const registerDuplicate = function() {
+  const duplicateOption = {
+    displayText: function() {
+      const selectedBlocksLength = blockSelection.size;
+      if (selectedBlocksLength <= 1) {
+        return Blockly.Msg['DUPLICATE_BLOCK'];
+      } else {
+        return Blockly.Msg['DUPLICATE_BLOCK'] + ' (' +
+        selectedBlocksLength + ')';
+      }
+    },
+    preconditionFn: function(scope) {
+      const block = scope.block;
+      if (!block.isInFlyout && block.isDeletable() && block.isMovable()) {
+        if (block.isDuplicatable()) {
+          return 'enabled';
+        }
+        return 'disabled';
+      }
+      return 'hidden';
+    },
+    callback: function(scope) {
+      const duplicatedBlocks = [];
+      const apply = function(block) {
+        if (block && duplicateOption.preconditionFn({
+          block: block,
+        }) === 'enabled' && !hasSelectedParent(block)) {
+          duplicatedBlocks.push(Blockly.clipboard.duplicate(block));
+        }
+      };
+      Blockly.Events.setGroup(true);
+      if (scope.block && blockSelection.size === 0) {
+        apply(scope.block);
+      }
+      blockSelection.forEach(function(id) {
+        const block = Blockly.getMainWorkspace().getBlockById(id);
+        if (block) {
+          apply(block);
+          block.pathObject.updateSelected(false);
+        }
+      });
+      Blockly.Events.setGroup(false);
+      blockSelection.clear();
+      duplicatedBlocks.forEach(function(block) {
+        if (block.id) {
+          blockSelection.add(block.id);
+          block.pathObject.updateSelected(true);
+        }
+      });
+    },
+    scopeType: Blockly.ContextMenuRegistry.ScopeType.BLOCK,
+    id: 'blockDuplicate',
+    weight: 1,
+  };
+  Blockly.ContextMenuRegistry.registry.register(duplicateOption);
+};
+
+/**
+ * Modification for context menu 'Comment' to be available for multiple blocks.
+ */
+const registerComment = function() {
+  const commentOption = {
+    displayText: function(scope) {
+      const selectedBlocksLength = blockSelection.size;
+      if (scope.block.getCommentIcon()) {
+        if (selectedBlocksLength <= 1) {
+          return Blockly.Msg['REMOVE_COMMENT'];
+        } else {
+          return Blockly.Msg['REMOVE_COMMENT'] + ' (' +
+          selectedBlocksLength + ')';
```

Comment:
Hmm technically `selectedBlocksLength` is inaccurate, because the code is only removing comments from blocks that already have them. It should probably search the `blockSelection` for the blocks that actually have comments. (similarly for adding comments)

Reply:
Now except for `Delete X Blocks`, I have changed all into counting the blocks which will actually take effects with respected action.